### ISSUE TITLE
RFC: Uses function

### ISSUE THREAD

**Issue by nunomaduro** [pestphp/pest#1]
**TL; DR**: This RFC it's about the `$this` binding of the given closures to the underlying PHPUnit test case class - It proposes the `uses` function that you can use to instruct Pest to use a specific class/trait at a file-level or directory-level.

![foo](https://user-images.githubusercontent.com/5457236/78892833-5c301b00-7a6a-11ea-9eb6-1da7ad6e64e7.png)

## Summary

By default, every given closure to globals functions of **Pest** is always bound to the current test case. As an example, on the code below, the `$this` is bound to `PhpUnit\Framework\TestCase`.

```php
it('has welcome page', function () {
    echo get_class($this); // Uses PhpUnit\Framework\TestCase
});
```

The goal of binding the closure to `$this` is giving access to underlying functionalities that may exist in current test case classes of the PHP Ecosystem - just like the Laravel's test case class.

To swap the underlying test case class, **Pest** proposes the `uses`function that can instruct **Pest** to use a specific class/trait a file level or at directory level. Here is an example with the Laravel Framework:

- Per directory directive:

```php
// Tests/Pest.php:
uses(TestCase::class, RefreshDatabase::class)->in('Features');

// Tests/Features/Example.php:
it('has welcome page', function () {
    echo get_class($this); // Uses Tests\TestCase + RefreshDatabase trait
});
```

- Per file directives:

Of course, you can also improve the granularity of this, using the `uses` function at file level:

```php
// Tests/Pest.php:
uses(TestCase::class)->in('Features');

// Tests/Features/Example.php:
uses(RefreshDatabase::class);

it('has welcome page', function () {
    echo get_class($this); // Uses Tests\TestCase + RefreshDatabase trait
});
```

One of the main goals of Pest is to provide seamless integration with **Laravel**. Everything should fell really natural and sexy. Let me know what do you think?

**Comment by deleugpn** [comment]
I haven't used or tested Pest at all, so if what I say makes no sense please ignore.

With Laravel routing system, I like to change my RouteServiceProvider to declare context-based route files. I have 5 route files and in the RouteServiceProvider I configure groups with a specific set of middleware.

This seems similar. Imagine

```
uses(TestCase::class, function() {
    it('has welcome page', function () {
        echo get_class($this); // Uses Tests\TestCase + RefreshDatabase::class
    });
}
```

The obvious problem being the increase of one level of indentation. Much like Laravel Routing system, you can move the entire group to it's own file to reduce the indentation.
That allows one file to declare all groups.

**Comment by musa11971** [comment]
To be honest, I had to look at it for at least 5 minutes before I could grasp what was going on, but now I think I understand how it works.  
  
> Let me know what do you think?  

It's *okay*. I don't per se like it, to be frank. The thing that confused me especially was:
```php
uses(TestCase::class, RefreshDatabase::class)->in('Features');
```
I doesn't make sense to me that you would put the class and trait next to each other like that.  
  
I came up with a different API that makes a bit more sense to me:

```php
// Tests/Pest.php:
testcase(CustomTestCase::class)->using(RefreshDatabase::class)->in('Features');

// Tests/Features/Example.php:
it('has welcome page', function () {
    echo get_class($this); // Uses Tests\CustomTestCase + RefreshDatabase trait
});
```

At file level:
```php
// Tests/Pest.php:
testcase(CustomTestCase::class)->in('Features');

// Tests/Features/Example.php:
using(RefreshDatabase::class);

it('has welcome page', function () {
    echo get_class($this); // Uses Tests\CustomTestCase + RefreshDatabase trait
});
```

I think this API is more expressive. By using the `testcase()` helper it is clear that you're declaring which testcase *class* to use. And the `using()` helper matches PHP's `use` keyword, that is of course used to implement traits into classes.  
The added benefit is that with this API you have classes and traits separated, rather than having them next to each other as arguments in your proposal.

Just my 2 cents! Not trying to be rude at all. I really like what you're trying to do @nunomaduro :)

**Comment by nunomaduro** [comment]
@musa11971 Thanks for your feedback, it seems that you are happy with the current way of doing things ( a global helper that instructs pest about things ) and you just want to change the api of that helper function.

**Comment by Parables** [comment]
Works when I use a trait
 throws this error when it is a class: `Test case X can not be used. The folder `ExampleTest.php` already uses the test case `TestCase`
